### PR TITLE
Fix lineage algo - RadioactiveDecayBase

### PR DESCRIPTION
**Pull request by cfuselli**
With this PR we fix some issues with the treatment of RadioactiveDecayBase (RDB) interactions. I realised that the results of simulating one gamma and simulating the gamma from Xe131 gave very different results. That is not good. Now they both give similar results. 

Specifically, we had to add special treatment for gammas emitted from excited nuclei that are treated by RDB library. 
Low energy gammas will not produce a gamma-phot/compt interaction in G4 just like hig energy gammas, but just some electron tracks and some gamma tracks will be produced. Now, we decide based on the secondaries what yield to assign to a certain cluster. The modifications match perfectly the distributions of Xe129 and Xe133 ( they were completely off before this PR ) and also match much better Pb212. 

## Details of changes: 

- If there is "[" in the particle type ( like Xe129[236.140] ) - we pass also the secondaries to the classification. We need them to understand if to give a gamma or beta yield.
- Added get_particle_secondaries function to get first order daughters of a particle
- classification with secondaries: if there are only gammas -> it's a beta yield compton. If there are also electrons, it's probably a low energy gamma, so gets a gamma yield. 
- lineage broken: brake if creaproc and edproc are both RDB. Intermediate step. 
- modify ion lineage broken. before: brake if lineage is 6, now checks if ion. 
- added: if lineage is a gamma yield, never brake anymore. We don't want to split photoabsorb. 
- added: 0.1cm threshold specially for brem. 

#### Xe129 - 39keV gamma
It makes some electron tracks and some gamma-phots. We want to keep them all together into a one gamma cluster. 

Xe129_39kev

#### Xe129 - 196keV gamma
It makes some electron tracks. Again, we want to keep them all together into a one gamma cluster. 

Xe129_196kev

#### Bi214 - 609keV gamma
It makes a real gamma. The gamma produces two compton clusters ( beta yield ) and one photoabsorbtion ( gamma yield ) 

BiPo_609kev

 ## Summary of clustering for different situations

Left: NEW clustering 
Right: OLD clustering

#### Gamma simulations 163keV in TPC
Expect just one gamma cluster, sometimes compton + gamma

Gamma_summary

#### Pb212
It's gammas+beta most of the times, with 238keV gamma

Pb212_summary

# Xe131
It's one 163keV gamma
Xe131_summary

# Xe129
It's two different gammas
Xe129_summary

# BiPo
there is gamma+beta and then alpha
BiPo_summary

### CODE REVIEW

File: fuse/plugins/micro_physics/lineage_cluster.py [hunk 1]
```diff
@@ -281,6 +300,24 @@ def get_parent(event_interactions, event_lineage, particle):
     return possible_parents[-1], possible_parents_lineages[-1]
 
 
+def get_all_particle_secondaries(event_interactions, particle):
+    """Returns all secondaries of the given particle.
+
+    It iterates over all interactions of the event and checks if the
+    parent id of the interaction is the same as the track id of the
+    particle, and continues down the tree until no more secondaries are
+    found or a RadioactiveDecayBase type interaction is found.
```
Comment (HenningSE) [inline fuse/plugins/micro_physics/lineage_cluster.py:309]: The docstring reads like the loop is supposed to check for RadioactiveDecayBase type interactions and then stop. There is no such think in the loop and I was wondering if the code needs to be updated or the docstring.

File: fuse/plugins/micro_physics/lineage_cluster.py [hunk 2]
```diff
@@ -295,10 +332,20 @@ def num_there(s):
...
-def classify_lineage(particle_interaction):
+def classify_lineage(particle_interaction, secondaries=None):
     """Function to classify a new lineage based on the particle and its parent
     information."""
 
+    # Check if we passed secondaries
+    if secondaries is not None:
+        for secondary in secondaries:
+            # there is a seconday that has type gamma and edproc phot,
+            # we classify this lineage as beta because it is most likely a compton scattering
```
Comment (HenningSE) [inline fuse/plugins/micro_physics/lineage_cluster.py:343]: This comment is also maybe outdated as the code does not check for gamma and phot edproc?

File: fuse/plugins/micro_physics/lineage_cluster.py [hunk 3]
```diff
@@ -362,7 +409,6 @@ def classify_lineage(particle_interaction):
         return NEST_NONE
 
 
-@njit()
```
Comment (HenningSE) [inline fuse/plugins/micro_physics/lineage_cluster.py:365]: Did you check how the speed of the code changes when removing numba? (Just want to make sure we do not slow done too much)

File: fuse/plugins/micro_physics/lineage_cluster.py [hunk 4]
```diff
@@ -372,18 +418,46 @@ def is_lineage_broken(
...
         return True
 
     # For gamma rays, check the distance between the parent and the particle
     if particle["type"] == "gamma":
 
+        if particle["creaproc"] == "phot" and particle["edproc"] == "phot":
+            # We do not want to split a photo absorption into two clusters
+            # The second photo absorption (that we see) could be x rays
+            # So, do not split if the distance is small
+            return False
```
Comment (HenningSE) [inline fuse/plugins/micro_physics/lineage_cluster.py:453]: Here you do not evaluate the distance between the energy deposits? Please update the comment or code :-)

File: fuse/plugins/micro_physics/lineage_cluster.py [hunk 5]
```diff
@@ -393,14 +467,13 @@ def is_lineage_broken(
...
-        if (particle["creaproc"] == "phot") and (particle["edproc"] == "phot"):
-            # we do not want to split a photo absorption into two clusters
-            # the second photo absorbtion (that we see) could be x rays
-            # so, do not split if the distance is small
-            return False
+        if particle["creaproc"] == "eBrem":
+            # we do not want to split a bremsstrahlung into two clusters
+            # if the distance is really small, it is most likely the same interaction
+            if distance < 0.1:  # cm
+                return False
```
Comment (HenningSE) [inline fuse/plugins/micro_physics/lineage_cluster.py:476]: I think we should put this into a URLConfig to keep track of it.

File: fuse/plugins/micro_physics/lineage_cluster.py (hunk 1 above)
Comment (cfuselli) [reply fuse/plugins/micro_physics/lineage_cluster.py:309]: Thanks Henning, I had it like that at the beginning but then realised I need only first generation secondaries. Updating docstring now :)

File: fuse/plugins/micro_physics/lineage_cluster.py (hunk 2 above)
Comment (cfuselli) [reply fuse/plugins/micro_physics/lineage_cluster.py:343]: Yes, also from previous version of the code. Thanks!

File: fuse/plugins/micro_physics/lineage_cluster.py (hunk 3 above)
Comment (cfuselli) [reply fuse/plugins/micro_physics/lineage_cluster.py:365]: Not checked yet.. but on small runs I saw no notable changes.

File: fuse/plugins/micro_physics/lineage_cluster.py (hunk 5 above)
Comment (cfuselli) [reply fuse/plugins/micro_physics/lineage_cluster.py:476]: Added :)

**Comment by cfuselli** [comment]
Btw I am still discussing with Micha about this PR and we are still not sure how we want to treat those cases.

**Comment by cfuselli** [comment]
With the latest commit we do a bit of "dietrofront" on the treatment of RDB decays. 
I will update the description of the PR asap. 

New treatment involves: 
- merge together all electrons coming from excited stated RDB decays, they are conversion electrons, and we give them a beta cluster. 
- keep doing the "if gamma break lineage". when we have both conversion electrons and gammas, we want to let the gammas travel and get a compton (beta yield) or photabs (gamma yield) cluster. 
- we remove the need to classify based on secondaries and all the cases can be handled with just a few lines, specifying to not break if there is a "[" in the parent and to assing an 8 if there is an "[" in the type.

**Pull request merged**